File: PythonSpark/Aggregations/Purchases/GroupedItemsAggregationJob.py
from pyspark.sql import DataFrame
from pyspark.sql.functions import column

from Aggregations.Purchases.PurchasesAggregationJob import PurchasesAggregationJob
from Utils.Parsing import count_data_frame_columns


class GroupedItemsAggregationJob(PurchasesAggregationJob):
    """
    This class calculates a number of times two items are bought together.
    Result is a dataframe with columns:
     item1 - first item
     item2 - second
     count - number item1 and item2 were bought together in specified locations, devices and time frame
    """

    def process(self, data: dict) -> DataFrame:
        filtered_purchases_with_locations = super(GroupedItemsAggregationJob, self).process(data)

        first_group = GroupedItemsAggregationJob.__select_item_group__(filtered_purchases_with_locations, "item1")
        second_group = GroupedItemsAggregationJob.__select_item_group__(filtered_purchases_with_locations, "item2")

        joined = first_group.join(second_group, "user_id", "ts")

        return count_data_frame_columns(joined, ["item1", "item2"])

    @staticmethod
    def __select_item_group__(data: DataFrame, new_item_id_column_name: str) -> DataFrame:
        return data.select("user_id", "ts", column("item_id").alias(new_item_id_column_name))

File: ScalaSpark/src/scala/Aggregations/Purchases/GroupedItemsAggregationJob.scala
package Aggregations.Purchases

import JobManagment.JobCompanion
import de.halcony.argparse.Parser
import org.apache.spark.sql.functions.column
import org.apache.spark.sql.{DataFrame, SparkSession}

/**
 * This class calculates a number of times two items are bought together
 * Result is a dataframe with columns:
 *   item1 - first item
 *   item2 - second
 *   count - number item1 and item2 were bought together in specified locations, devices and time frame
 */
class GroupedItemsAggregationJob(arguments: Map[String, String], spark: SparkSession) extends PurchasesAggregationJob(arguments, spark) {
  override def process(data: Map[String, DataFrame]): DataFrame = {
    val filteredPurchasesWithLocations = super.process(data)

    val firstGroup = selectItemGroupFromPurchases(filteredPurchasesWithLocations, "item1")

    val secondGroup = selectItemGroupFromPurchases(filteredPurchasesWithLocations, "item2")

    val joined = firstGroup.join(secondGroup, "ts", "user_id")

    val filteredOutGroupsWithSameItems = joined.filter(column("item1") > column("item2"))

    filteredOutGroupsWithSameItems.groupBy("item1", "item2").count()
  }

  def selectItemGroupFromPurchases(purchases: DataFrame, newItemIdColumnName: String): DataFrame = {
    purchases.select(column("ts"), column("user_id"), column("item_id").as(newItemIdColumnName))
  }
}

object GroupedItemsAggregationJob extends JobCompanion {
  override def parser(): Parser = PurchasesAggregationJob.parser()
}